packages/router/src/grpc/view-protocol-server/broadcast-transaction.test.ts

```diff
@@ -0,0 +1,483 @@
+import { beforeEach, describe, expect, Mock, test, vi } from 'vitest';
+import {
+  BroadcastTransactionRequest,
+  BroadcastTransactionResponse,
+  TransactionInfo,
+} from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/view/v1alpha1/view_pb';
+import { createContextValues, createHandlerContext, HandlerContext } from '@connectrpc/connect';
+import { ViewProtocolService } from '@buf/penumbra-zone_penumbra.connectrpc_es/penumbra/view/v1alpha1/view_connect';
+import { servicesCtx } from '../../ctx';
+import { Transaction } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/transaction/v1alpha1/transaction_pb';
+import { broadcastTransaction } from './broadcast-transaction';
+import type { Services } from '@penumbra-zone/services';
+import { TransactionId } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/txhash/v1alpha1/txhash_pb';
+
+const mockSha256 = vi.hoisted(() => vi.fn());
+vi.mock('@penumbra-zone/crypto-web', () => ({
+  sha256Hash: mockSha256,
+}));
```

Comment:
I'm not sure if it's necessary, but the CI-CD pipeline was throwing a 'crypto is not defined' error, although locally everything worked without this mock.

Reply:
You can't just do?:

```typescript
vi.mock('@penumbra-zone/crypto-web', () => ({
  sha256Hash: vi.fn(),
}));
```

Reply:
`sha256` must return the correct hash, otherwise we will always get the error 'broadcast transaction id disagrees'.

```typescript
const id = new TransactionId({ inner: await sha256Hash(req.transaction.toBinary()) });

  const broadcastId = await tendermint.broadcastTx(req.transaction);
  if (!id.equals(broadcastId)) {
    console.error('broadcast transaction id disagrees', id, broadcastId);
    throw new Error(
      `broadcast transaction id disagrees: expected ${uint8ArrayToHex(id.inner)} != tendermint ${uint8ArrayToHex(broadcastId.inner)}`,
    );
  }
```